### ISSUE TITLE
Any io_uring performance tests after all of the work that has been done?

### ISSUE THREAD

**Issue by Qix-** [netty/netty#10622]
I'm not an active (anymore) Netty user but Netty is the largest project I've seen so far that has incorporated io_uring support.

However, several individuals have called into question the _actual_ increases in efficiency, as seen over at https://github.com/axboe/liburing/issues/189 and https://github.com/frevib/io_uring-echo-server/issues/8. Simply put, io_uring appears to be slower than epoll in many of our tests; certainly not the 68% increase in performance that has been claimed by the benchmark maintainer as well as the inventor of io_uring.

I'm curious to know if any performance differences have been seen with applications that use Netty using the io_uring implementation as opposed to epoll or the like. Since one of the claims was that the benchmark was not complex enough (which I don't _personally_ find to be a valid criticism given the circumstances) I figured I'd open an issue here since if Netty isn't complex enough then I don't know what is 🙃 

Thanks for any information.

**Comment by franz1981** [comment]
Hi @Qix- ,

I'm going to answer you by using your own last comment on https://github.com/axboe/liburing/issues/189#issuecomment-701119442:

> Part of the scientific process is to be able to reproduce claims like those being made (60%+ increase in performance over epoll, apparently that was 99% at one point but bugs were found). These are metrics that @axboe has not refuted and has seemingly even confirmed, especially through promoting it on twitter and urging others to buy in (e.g. Netty).

Part of a complete scientific process, if we're talking about performance engineering and not the "classic" SW engineering, is to understand and investigate what's happening on the test rig in order to correctly interpret results: similar to physics or any other empirical science, we just don't need "benchmarks" but "benchmarks and clean reproducible environment + any means to understand if the environments is kept clean and reliable to have valid results"; this isn't something that come just by adopting the correct benchmark.

Said that, Netty is a OSS project (please remember that I'm Netty user like you) that's used for many different use cases: I'm sure you have your own use cases to care about (mine is a messaging broker using Netty) so the idea is that you can publish your numbers on https://github.com/netty/netty/issues and we can freely speak/try together to understand what's happened and why you got bad/good/similar performance, with all the data needed to understand it.

> I'm curious to know if any performance differences have been seen with applications that use Netty using the io_uring implementation as opposed to epoll or the like

If I'll run the messaging test suite performance benchmark soon and I got some reliable results I will publish here, but...
messaging brokers are complex beasts and most of the time the networking activity is just part of the story: memory consumption, cpu utilization and other factors come in place that can make comparisons not easy.
Even running a RUL test (responsiveness under load), correctly addressing [coordinated omission](https://www.youtube.com/watch?v=lJ8ydIuPFeU&ab_channel=StrangeLoop) but using the wrong target throughput could be misleading if most of the time for both epoll and io_uring are awaking threads; just to give you an example of scenario I'm concerned about.

**Comment by Qix-** [comment]
> this isn't something that come just by adopting the correct benchmark.

I'm not sure what you mean here. The inventors of io_uring claim "free performance" over epoll. I'm asking if Netty has seen any such "free" performance gains by switching to the io_uring implementation.

> Share them in the appropriate place and we can look at the results together to understand what's going on :)

We did. Please look at the thread again. Following the exact instructions from the performance benchmarks, they are claiming 68% increase (they were claiming 99% increase prior to realizing there were some bugs) of "free performance", where we are not seeing anything _near_ the sort.

I decided to see if any real-world applications with Netty have switched to using an io_uring implementation and seen any characteristics of a performance increase.

Further, my comment about being scientific was more of a comment on the Twitter conversations happening. Not sure why that's relevant here, in this thread.

**Comment by franz1981** [comment]
> The inventors of io_uring claim "free performance" over epol

Yes, but I'm not sure any of the Netty dev is the io_uring inventor, hence I'm not sure makes sense to post a comment about their claims here.

>We did. Please look at the thread again

Yes, but it's not running on Netty, so I'm not sure this is the *appropriate* place 
I see the usage of the word *they* pretty evident in your comment, but none of such *they* is here on this project: I still think any of your previous comment isn't relevant to start a friendly investigation on any Netty performance issue (or claim, that has never been done, as you've rightly noticed :) )

> Further, my comment about being scientific was more of a comment on the Twitter conversations happening. Not sure why that's relevant here, in this thread.

Given that we're not talking about claims, but about real data; the comment was an invitation to share something appropriate for the place the comment has been written: a Netty issue.
If you come with some Netty test, feel free to share it (again) ;)

**Comment by Qix-** [comment]
@franz1981 Are you here to provide data or argue with me? I'm not looking for fights.

If you check the slew of recent io_uring pull requests, you can see that Netty has an interest in supporting the new set of syscalls. I'm simply asking if there are benchmarks available for Netty applications to compare the differences, to see how io_uring impacts performance.

> Netty performance issue

Huh? I've not made _any_ claims about Netty's performance, negative or positive. I'm looking into the performance claims of the io_uring subsystem. Not Netty specifically.

> Given that we're not talking about claims, but about real data; the comment was an invitation to share something appropriate for the place the comment has been written: a Netty issue.

Please stop policing my questions. That's all it was - a question. Please stop assuming malice. There's nothing wrong with the issue I've opened here.

**Comment by franz1981** [comment]
> Are you here to provide information or argue with me? I'm not looking for fights.

Good to hear it, me too :)

Despite that; I personally think isn't appropriate to propose claims of others (not belonging to this project) as the engine to guide your questions: I cannot say if is a cultural difference here or just personal taste, but I'm sure it could be avoided while exposing your problem; it's just not friendly with both Netty devs and io_uring ones.

> I'm simply asking if there are benchmarks available for Netty applications to compare the differences, to see how io_uring impacts performance.

AFAIK nope, as said, but some tests has been run using simple Echo (Rust and not only) tests to compare different builds of Netty while developing its support for io_uring and not to compare it with epoll.
I'm going to publish here some results for my use case if they will be relevant for Netty io_uring support, at some point.

> Huh? I've not made any claims about Netty's performance, negative or positive. I'm looking into the performance claims of the io_uring subsystem. Not Netty specifically.

If you're searching for io_uring performance claims I'm still not sure this is the right place to search for: Netty interact a lot with native layers using JNI (while using io_uring much less, under heavy load) so this add a variable that cannot be ignored in the results, especially if you're not a Netty user anymore.

> Please stop policing my questions. That's all it was - a question. Please stop assuming malice. There's nothing wrong with the issue I've opened here.

I will left other committers/devs to answer your question, hopefully with some benchmark data results: I'm not assuming anything, but just trying to understand how I can help you on this issue.

**Comment by franz1981** [comment]
@Qix- 
I've just changed the label of your question to help others to find this and possibly paste any results :+1:

**Comment by doom369** [comment]
@Qix- netty doesn't ship io_uring yet in it's builds, also it's actively developed right now. So it's very unlikely we gonna see some real-world benchmarks soon. I'm personally eager to try it as well.

**Comment by normanmaurer** [comment]
@Qix- all I can say for now is that we see some significant performance improvements once the connection count increases. I verified this by two different tools ... `tcpkali` and the `rust_echo_bench`.  We talking about 100% perf win when using `tcpkali` as client and even more when using `rust_echo_bench` here. That said this is for a throughput benchmark and I still need to do some latency benchmarks. 

Just to give you some idea...

epoll based
```
./tcpkali  -m xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx  --connect-rate=200 -c 500 -T 30 127.0.0.1:8088
Destination: [127.0.0.1]:8088
Interface lo address [127.0.0.1]:0
Using interface lo to connect to [127.0.0.1]:8088
Ramped up to 500 connections.
Total data sent:     35254.6 MiB (36967084668 bytes)
Total data received: 35725.3 MiB (37460714661 bytes)
Bandwidth per channel: 39.688⇅ Mbps (4961.0 kBps)
Aggregate bandwidth: 9987.847↓, 9856.234↑ Mbps
Packet rate estimate: 861701.3↓, 851472.0↑ (11↓, 43↑ TCP MSS/op)
Test duration: 30.005 s.

Running `target/release/echo_bench --address 'localhost:8088' --number 500 --duration 30 --length 128`
Benchmarking: localhost:8088
500 clients, running 128 bytes, 30 sec.

Speed: 80820 request/sec, 80820 response/sec
Requests: 2424626
Responses: 2424625
```

io_uring based:
```
root@ubuntu-dev:/home/norman/Apps/tcpkali/bin# ./tcpkali  -m xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx  --connect-rate=200 -c 500 -T 30 127.0.0.1:8081
Destination: [127.0.0.1]:8081
Interface lo address [127.0.0.1]:0
Using interface lo to connect to [127.0.0.1]:8081
Ramped up to 500 connections.
Total data sent:     74653.1 MiB (78279409944 bytes)
Total data received: 75180.9 MiB (78832885977 bytes)
Bandwidth per channel: 83.793⇅ Mbps (10474.1 kBps)
Aggregate bandwidth: 21021.980↓, 20874.387↑ Mbps
Packet rate estimate: 1932417.0↓, 1793964.8↑ (11↓, 39↑ TCP MSS/op)
Test duration: 30.0002 s.

Running `target/release/echo_bench --address 'localhost:8081' --number 500 --duration 30 --length 128`
Benchmarking: localhost:8081
500 clients, running 128 bytes, 30 sec.

Speed: 267371 request/sec, 267370 response/sec
Requests: 8021137
Responses: 8021128
```

I know this is on localhost and should be redone in labs but I just wanted to share something quickly with you. This is on kernel 5.9-rc4.

This benchmark is basically sending data and expect the data to be echoed back. The server itself uses 1 thread to handle all of this.

**Comment by Qix-** [comment]
@normanmaurer Thank you, this is exactly what I was hoping for. Gives me a jumping off point to look into it myself.

**Comment by normanmaurer** [comment]
@Qix- no problem... I hope we can improve these further in the future

**Comment by Qix-** [comment]
It's hard to stomache claims that a magical new low-level feature gives us more than a few percentages of speedup. That's why the heavy skepticism. We've seen this before in the kernel world and we'll see it again. I think healthy skepticism of a new, experimental feature is just that - healthy. :)

**Comment by pcarrier** [comment]
@Qix- A small change to the characteristics of a system can have a huge impact on its consumers. For example, in Linux, exposing a vDSO allowed processes to read the current system time through roughly a register (with the right clock source, eg no virtualization) instead of a system call; because the previous time was mostly due to the overhead of a system call and not the actual work, it can absolutely be orders of magnitude faster in real workloads.

This is such a change. It won't affect all workloads nearly as much, but for any server working on small requests across many connections (eg a memcached server or MQTT broker), in the absence of pipelining (such that at most one request is serviceable on any given socket), the overhead of at least two system calls per request can be considerable compared to the actual work. That's driven a lot of the interest in userspace networking stacks; io_uring can deliver similar performance benefits, without reinventing the wheel we've been rounding for decades.

**Comment by franz1981** [comment]
@pcarrier @Qix- to not mention that JNI cost: that's why saving both can change dramatically results and that's why applying io_uring on Netty is the same then, for example, doing it on Rust or on C.

**Comment by njhill** [comment]
@normanmaurer the results looks great, that's awesome. Out of curiosity how do the numbers look for netty-native ET epoll rather than jdk epoll?

Also are there any places in netty epoll impl where we could also use Unsafe to avoid some JNI, or do we have to cross the boundary regardless everywhere that JNI that's used in that case? guessing no or you would have already done that...

Would just be interesting to minimze those variables when comparing straight io_uring versus epoll perf.

**Comment by LifeIsStrange** [comment]
It would be interesting to re-run those benchmarcks on a recent kernel and with the many improvements that the Netty io_uring integration has received.
Most importantly, the benchmarks should be run with Linux mitigations=off as it can have a huge differentiated impact on epoll Vs io_uring.

**Comment by normanmaurer** [comment]
Honestly I am not sure if benchmarks should set mitigations=off as this is not really something you should do in real world due security concerns

**Comment by LifeIsStrange** [comment]
@normanmaurer the best would be to benchmarck with and without. 
The security concerns of speculative executions attack are theoretical and are extremely, extremely hard to do and are extremely slow as the bitrate of extraction of priviliegied data is extremely low+noisy. 
It is an open question whether in practice someone can leverage it and my unpopular opinion is that someone would win MUCH more security wise by enabling opt-in security options (e.g grsecurity, fine-tuning SElinux, doing sandboxing, various hardening flags e.g PIE, etc) without loosing as much performance.

I'm not advising for companies in general to disable it but it is a technical decision that can make sense if you have performance critical workloads and that you carefully think about the surface attack (and the other low hanging fruits in security). Moreover, the mitigations do *not* make sense for computers that are not connected to internet or that only have trusted intranets.
Hence why benchmarck without the mitigations would be a useful information

**Comment by ran-su** [comment]
> all I can say for now is that we see some significant performance improvements once the connection count increases. I verified this by two different tools ... `tcpkali` and the `rust_echo_bench`. We talking about 100% perf win when using `tcpkali` as client and even more when using `rust_echo_bench` here. That said this is for a throughput benchmark and I still need to do some latency benchmarks.
> 
> Just to give you some idea...
> 
> epoll based
> 
> ```
> ./tcpkali  -m xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx  --connect-rate=200 -c 500 -T 30 127.0.0.1:8088
> Destination: [127.0.0.1]:8088
> Interface lo address [127.0.0.1]:0
> Using interface lo to connect to [127.0.0.1]:8088
> Ramped up to 500 connections.
> Total data sent:     35254.6 MiB (36967084668 bytes)
> Total data received: 35725.3 MiB (37460714661 bytes)
> Bandwidth per channel: 39.688⇅ Mbps (4961.0 kBps)
> Aggregate bandwidth: 9987.847↓, 9856.234↑ Mbps
> Packet rate estimate: 861701.3↓, 851472.0↑ (11↓, 43↑ TCP MSS/op)
> Test duration: 30.005 s.
> 
> Running `target/release/echo_bench --address 'localhost:8088' --number 500 --duration 30 --length 128`
> Benchmarking: localhost:8088
> 500 clients, running 128 bytes, 30 sec.
> 
> Speed: 80820 request/sec, 80820 response/sec
> Requests: 2424626
> Responses: 2424625
> ```
> 
> io_uring based:
> 
> ```
> root@ubuntu-dev:/home/norman/Apps/tcpkali/bin# ./tcpkali  -m xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx  --connect-rate=200 -c 500 -T 30 127.0.0.1:8081
> Destination: [127.0.0.1]:8081
> Interface lo address [127.0.0.1]:0
> Using interface lo to connect to [127.0.0.1]:8081
> Ramped up to 500 connections.
> Total data sent:     74653.1 MiB (78279409944 bytes)
> Total data received: 75180.9 MiB (78832885977 bytes)
> Bandwidth per channel: 83.793⇅ Mbps (10474.1 kBps)
> Aggregate bandwidth: 21021.980↓, 20874.387↑ Mbps
> Packet rate estimate: 1932417.0↓, 1793964.8↑ (11↓, 39↑ TCP MSS/op)
> Test duration: 30.0002 s.
> 
> Running `target/release/echo_bench --address 'localhost:8081' --number 500 --duration 30 --length 128`
> Benchmarking: localhost:8081
> 500 clients, running 128 bytes, 30 sec.
> 
> Speed: 267371 request/sec, 267370 response/sec
> Requests: 8021137
> Responses: 8021128
> ```
> 
> I know this is on localhost and should be redone in labs but I just wanted to share something quickly with you. This is on kernel 5.9-rc4.
> 
> This benchmark is basically sending data and expect the data to be echoed back. The server itself uses 1 thread to handle all of this.
> 
> ```java
> @Sharable
> public class EchoServerHandler extends ChannelInboundHandlerAdapter {
> 
>     @Override
>     public void channelRead(ChannelHandlerContext ctx, Object msg) {
>         ctx.write(msg);
>     }
> 
>     @Override
>     public void channelReadComplete(ChannelHandlerContext ctx) {
>         ctx.flush();
>     }
> 
>     @Override
>     public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
>         // Close the connection when an exception is raised.
>         ctx.close();
>     }
> 
>     @Override
>     public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exception {
>         // Ensure we are not writing to fast by stop reading if we can not flush out data fast enough.
>         if (ctx.channel().isWritable()) {
>             ctx.channel().config().setAutoRead(true);
>         } else {
>             ctx.flush();
>             if (!ctx.channel().isWritable()) {
>                 ctx.channel().config().setAutoRead(false);
>             }
>         }
>     }
> }
> 
> // This is using io_uring
> public class EchoIOUringServer {
>     private static final int PORT = Integer.parseInt(System.getProperty("port", "8081"));
> 
>     public static void main(String []args) {
>         EventLoopGroup group = new IOUringEventLoopGroup(1);
>         final EchoServerHandler serverHandler = new EchoServerHandler();
>         try {
>             ServerBootstrap b = new ServerBootstrap();
>             b.group(bossGroup, workerGroup)
>                     .option(ChannelOption.SO_REUSEADDR, true)
>                     .channel(IOUringServerSocketChannel.class)
>                     .childHandler(new ChannelInitializer<SocketChannel>() {
>                         @Override
>                         public void initChannel(SocketChannel ch) throws Exception {
>                             ChannelPipeline p = ch.pipeline();
>                             //p.addLast(new LoggingHandler(LogLevel.INFO));
>                             p.addLast(serverHandler);
>                         }
>                     });
> 
>             // Start the server.
>             ChannelFuture f = b.bind(PORT).sync();
> 
>             // Wait until the server socket is closed.
>             f.channel().closeFuture().sync();
>         } catch (InterruptedException e) {
>             e.printStackTrace();
>         } finally {
>             // Shut down all event loops to terminate all threads.
>             group.shutdownGracefully();
>         }
>     }
> }
> 
> // This is using epoll / syscalls 
> public class EchoNioServer {
>     private static final int PORT = Integer.parseInt(System.getProperty("port", "8088"));
> 
>     public static void main(String []args) {
>         EventLoopGroup group = new NioEventLoopGroup(1);
>         final EchoServerHandler serverHandler = new EchoServerHandler();
>         try {
>             ServerBootstrap b = new ServerBootstrap();
>             b.group(bossGroup, workerGroup)
>                     .option(ChannelOption.SO_REUSEADDR, true)
>                     .channel(NioServerSocketChannel.class)
>                     .childHandler(new ChannelInitializer<SocketChannel>() {
>                         @Override
>                         public void initChannel(SocketChannel ch) throws Exception {
>                             ChannelPipeline p = ch.pipeline();
>                             //p.addLast(new LoggingHandler(LogLevel.INFO));
>                             p.addLast(serverHandler);
>                         }
>                     });
> 
>             // Start the server.
>             ChannelFuture f = b.bind(PORT).sync();
> 
>             // Wait until the server socket is closed.
>             f.channel().closeFuture().sync();
>         } catch (InterruptedException e) {
>             e.printStackTrace();
>         } finally {
>             // Shut down all event loops to terminate all threads.
>             group.shutdownGracefully();
>         }
>     }
> }
> ```

Hi @normanmaurer,

I tried your benchmark and some gRPC benchmarks recently on a machine running Debian (kernel 6.7.12). The numbers I got are significantly different from your results (**io_uring is slower**). I understand io_uring and Netty's io_uring implementation changed a lot in the past few years. Do you think it is possible that you could re-do some benchmarks on your side and share some more recent results with us?

The results I got:
gRPC java epoll vs io_uring: localhost, 10K payload, plaintext, 10 channels.
NIO
```
Benchmark                                                     (payloadSize)  (rpc)  (securityLevel)  (transport)    Mode      Cnt     Score   Error  Units
TransportBenchmark.unaryCallLatency                                   10240   GRPC             NONE     NETTY_ALL  sample  2679699   186.400 ± 0.083  us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p0.00            10240   GRPC             NONE     NETTY_ALL  sample             80.896          us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p0.50            10240   GRPC             NONE     NETTY_ALL  sample            183.040          us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p0.90            10240   GRPC             NONE     NETTY_ALL  sample            224.768          us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p0.95            10240   GRPC             NONE     NETTY_ALL  sample            239.360          us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p0.99            10240   GRPC             NONE     NETTY_ALL  sample            274.944          us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p0.999           10240   GRPC             NONE     NETTY_ALL  sample            393.216          us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p0.9999          10240   GRPC             NONE     NETTY_ALL  sample           1685.565          us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p1.00            10240   GRPC             NONE     NETTY_ALL  sample           2187.264          us/op
```
IO_URING
```
Benchmark                                                     (payloadSize)  (rpc)  (securityLevel)  (transport)    Mode      Cnt     Score   Error  Units
TransportBenchmark.unaryCallLatency                                   10240   GRPC             NONE     NETTY_IO_URING  sample  2537658   196.842 ± 0.092  us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p0.00            10240   GRPC             NONE     NETTY_IO_URING  sample             86.400          us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p0.50            10240   GRPC             NONE     NETTY_IO_URING  sample            193.024          us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p0.90            10240   GRPC             NONE     NETTY_IO_URING  sample            238.848          us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p0.95            10240   GRPC             NONE     NETTY_IO_URING  sample            254.464          us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p0.99            10240   GRPC             NONE     NETTY_IO_URING  sample            289.792          us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p0.999           10240   GRPC             NONE     NETTY_IO_URING  sample            427.520          us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p0.9999          10240   GRPC             NONE     NETTY_IO_URING  sample           1658.880          us/op
TransportBenchmark.unaryCallLatency:unaryCallLatency·p1.00            10240   GRPC             NONE     NETTY_IO_URING  sample           8118.272          us/op
```

tcpkali:
tcpkali  -m {some 128B payload}  --connect-rate=200 -c 500 -T 30 127.0.0.1:8088

NIO:
```
Destination: [127.0.0.1]:8088
Interface lo address [127.0.0.1]:0
Using interface lo to connect to [127.0.0.1]:8088
Ramped up to 500 connections.
Total data sent:     77540.8 MiB (81307435008 bytes)
Total data received: 76538.1 MiB (80256030976 bytes)
Bandwidth per channel: 86.161⇅ Mbps (10770.1 kBps)
Aggregate bandwidth: 21400.087↓, 21680.442↑ Mbps
Packet rate estimate: 1959483.9↓, 1860847.2↑ (12↓, 45↑ TCP MSS/op)
Test duration: 30.0021 s.
```

IO_URING:
```
Destination: [127.0.0.1]:8088
Interface lo address [127.0.0.1]:0
Using interface lo to connect to [127.0.0.1]:8088
Ramped up to 500 connections.
Total data sent:     60919.4 MiB (63878594560 bytes)
Total data received: 60833.3 MiB (63788376064 bytes)
Bandwidth per channel: 68.080⇅ Mbps (8509.9 kBps)
Aggregate bandwidth: 17007.867↓, 17031.922↑ Mbps
Packet rate estimate: 1557492.9↓, 1461861.6↑ (12↓, 45↑ TCP MSS/op)
Test duration: 30.0042 s.
```

**Comment by normanmaurer** [comment]
@ran-su yes I can... that said it might take a bit of time as I just returned from vacation and have quite some backlog to process

**Comment by franz1981** [comment]
@ran-su depending on the number of event loop and the number of connections per event loop there should be (it was, before) the chance that syscalls are turned asynchronous with io_uring, and based on the kernel version this can give a much worse performance than Netty. Additionally, the number of event loops per CPUs (real cores) can severely affect io_uring performance. Do you have the option to use a sync profiler and share some data? Would be nice to have CPU usage and the exact configuration.

**Comment by ran-su** [comment]
> @ran-su depending on the number of event loop and the number of connections per event loop there should be (it was, before) the chance that syscalls are turned asynchronous with io_uring, and based on the kernel version this can give a much worse performance than NIO. Additionally, the number of event loops per CPUs (real cores) can severely affect io_uring performance. Do you have the option to use a sync profiler and share some data? Would be nice to have CPU usage and the exact configuration. I am still in vacation as well so my answers will happen with a "best effort" mode

For tcpkali benchmarks I used Norman's code above with changes that fix the boss and worker ELG references. All the ELGs are all single threaded.

For the above gRPC benchmarks I used all default config (I just realized the number of threads in ELG were different, gRPC by default create ELG with `number of core` threads, but Netty's new XxxEventLoopGroup() by default creates `2 * number of core` threads). However, after I tried with let them both use `2 * # of cores` ELGs, the new results were very similar to the gRPC results above.

The benchmarks were executed on an AMD Zen3 12 core CPU with hyperthread (NettyRuntime.availableProcessors() = 24) and 128G RAM machine.

I haven't got time to do any profiling yet. I was think I should check with the Netty community first before continue digging in case other people saw this kind of result before. I'll try to find some time to do some profiling and update here once I have new data.
